src/drivers/imx/edma.c

```diff
@@ -3,115 +3,520 @@
+	 */
+
+	/* TODO Support more advanced non-SG configurations */
+
+	rc = edma_validate_nonsg_config(sgelems, soff, doff);
+	if (rc < 0)
+		return rc;
+
+	sbase = sgelems->elems[0].src;
+	dbase = sgelems->elems[0].dest;
+	total_size = 2 * sgelems->elems[0].size;
+	/* TODO more flexible elem_count and elem_size
+	 * calculations
+	 */
+	elem_count = 2;
+	elem_size = total_size / elem_count;
+
+	rc = edma_compute_attr(src_width, dest_width);
+	if (rc < 0)
+		return rc;
+
+	/* Configure the in-hardware TCD */
+	dma_chan_reg_write(channel, EDMA_TCD_SADDR, sbase);
+	dma_chan_reg_write16(channel, EDMA_TCD_SOFF, soff);
+	dma_chan_reg_write16(channel, EDMA_TCD_ATTR, rc);
+	dma_chan_reg_write(channel, EDMA_TCD_NBYTES, elem_size);
+	dma_chan_reg_write(channel, EDMA_TCD_SLAST, -total_size * SGN(soff));
+	dma_chan_reg_write(channel, EDMA_TCD_DADDR, dbase);
+	dma_chan_reg_write16(channel, EDMA_TCD_DOFF, doff);
+	dma_chan_reg_write16(channel, EDMA_TCD_CITER, elem_count);
+	dma_chan_reg_write(channel, EDMA_TCD_DLAST_SGA,
+			   -total_size * SGN(doff));
+	dma_chan_reg_write16(channel, EDMA_TCD_BITER, elem_count);
+	dma_chan_reg_write16(channel, EDMA_TCD_CSR, 0);
+
+	channel->status = COMP_STATE_PREPARE;
 	return 0;
 }
 
 /* set the DMA channel configuration, source/target address, buffer sizes */
 static int edma_set_config(struct dma_chan_data *channel,
 			   struct dma_sg_config *config)
 {
-	return 0;
+	int handshake, irq, i;
+	uint16_t soff = 0;
+	uint16_t doff = 0;
+
+	/* We may need to pass some data through the handshake in the
+	 * future, so we just have them here, even if unused for now
+	 */
+	(void)handshake;
+	(void)irq;
+
+	tracev_edma("EDMA: set config");
+	tracev_edma("EDMA: source width %d dest width %d burst elems %d",
+		    config->src_width, config->dest_width, config->burst_elems);
+
+	switch (config->direction) {
+	case DMA_DIR_MEM_TO_DEV:
+		soff = config->src_width;
+		doff = 0;
+		handshake = config->dest_dev;
+		break;
+	case DMA_DIR_DEV_TO_MEM:
+		soff = 0;
+		doff = config->dest_width;
+		handshake = config->src_dev;
+		break;
+	default:
+		trace_edma_error("edma_set_config() unsupported config direction");
+		return -EINVAL;
+	}
+
+	tracev_edma("EDMA: SOFF = %d DOFF = %d", soff, doff);
+	tracev_edma("EDMA: src dev %d dest dev %d", config->src_dev,
+		    config->dest_dev);
+	tracev_edma("EDMA: cyclic = %d", config->cyclic);
+	if (!config->cyclic) {
+		trace_edma_error("EDMA: Only cyclic configurations are supported!");
+		return -EINVAL;
+	}
+	if (config->scatter) {
+		trace_edma_error("EDMA: scatter enabled, that is not supported for now!");
+		return -EINVAL;
+	}
+
+	/* IRQ is actually handled by SOF itself so I don't need to
+	 * consider it myself.
+	 */
+	tracev_edma("EDMA: irq_disabled = %d", config->irq_disabled);
+
+	tracev_edma("EDMA: %d elements", config->elem_array.count);
+	for (i = 0; i < config->elem_array.count; i++)
+		tracev_edma("EDMA: elem %d src 0x%08x -> dst 0x%08x size 0x%x bytes",
```

Comment:
the macro, when defined as empty will produce this code

 for (i = 0; i < config->elem_array.count; i++)
                                         ;
 return edma_setup_tcd(channel, soff, doff, config->cyclic,
         config->scatter, config->irq_disabled,
         &config->elem_array, config->src_width,
         config->dest_width);
i'd wrap the for with
#if CONFIG_TRACE

Reply:
I think that isn't necessary, losing a few cycles for the iteration itself when the verbose tracing is disabled may not be a real issue since this one is done in set_config(), not during actual playback. Will keep in mind though.

Reply:
Moved them to my own branch (I still need those prints for debugging).

Reply:
Oh now I did that.

Reply:
Thanks @paulstelian97